spi-proposal.md

```diff
@@ -0,0 +1,200 @@
+the last byte sent. However, the user can change the active CS pin state (HIGH
+to LOW rather than the default LOW to HIGH behavior) or even specify if the
+pin should be toggled between every byte vs the beginning and end of every
+sequence of bytes.
+
+A **sequence** byte is provide to allow breaking up a transfer into multiple
+packets or even enabling a continuous data stream (provided HW doesn't have
+tight timing requirements).
+
+There are 3 ways to send and receive data from the SPI slave device:
+
+1. Transfer - for each byte sent a byte is received
+2. Write - only write data (ignore any data retuned by the slave device)
+3. Read - only read data, writing 0 for each byte to be read
+
+
+## Protocol details
+
+### Begin
+
+Call once to initialize SPI hardware. Can optionally be set internally upon
+receiving first `SPI_CONFIG_DEVICE` message.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_BEGIN
+3:  channel (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### Config
+
+Send once for each SPI device, specifying a new deviceId for each device
+(beginning from 0).
+
+Bit order: LSBFIRST = 0, MSBFIRST = 1
+
+Data modes:
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+```
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_CONFIG_DEVICE
+3:  deviceId | channel (bits 2-6: deviceId, bits 0-1: channel)
+4:  bitOrder (bit 0) | dataMode (bits 1-2, values 0-3) see mode column in table above
+5:  clockSpeed (bits 0 - 6)
+6:  clockSpeed (bits 7 - 14)
+7:  clockSpeed (bits 15 - 21)
+8:  clockSpeed (bits 22 - 28)
+9:  clockSpeed (bits 29 - 32)
+10: csPin (0-127) [optional] use csPin to identify individual devices
+11: csOptions (used only if csPin is set):
+      csActive (bit 1) default = 0 (0 = LOW -> HIGH, 1 = HIGH -> LOW)
+      csToggle (bit 2) default = 0 (0 = toggle after last byte, 1 = toggle between bytes)
+12: END_SYSEX
+```
+
+### Transfer
+
+**Read a bybe for each byte written.**
+
+Send an array of data to transfer to the SPI slave device, expecting a byte
+returned for each byte written. Respond with a `SPI_REPLY` message.
+
+**sequence**
+
+Use the sequence byte to specify if this is a one time transfer or if it is part
+of a sequence of transfers, or an ongoing stream of transfers. Define as
+follows:
+
+```
+0 = single transfer (no sequence)
+
+1 = start sequence (1st part of sequence)
+2 = 2nd part of sequence
+3 = 3rd part of sequence, etc...
+... up to 100, after 100, wrap back to 1
+
+127 = end sequence
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_TRANSFER_REQUEST
+3:  deviceId | channel (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence (see note on sequence above)
+5:  numBytes to read and write
```

Comment:
any reason we'd need to write a different number of bytes than read? Perhaps if you write to a register and that has a multi-byte response. If that is a case, we'd need a separate numByteToWrite and numBytesToRead.